### ISSUE TITLE
Performance implications of using JinjaX components

### ISSUE THREAD

**Issue by MadCatX** [Molecular-Biophysics-Database/mbdb-app#27]
TL;DR
It is clear that JinjaX is not indented for use cases where performance is of any concern whatsoever. Its use in `mbdb-app` must be avoided until such time when JinjaX's performance is comparable to plain Jinja.

When I mostly finished the rewrite of the original record view code to make use of JinjaX components, I noticed severe performance degradation. A single page with a randomly generated MST record took 12 (twelve) seconds to render, a result that is clearly problematic. While the randomly generated records are quite large and I have been testing the JinjaX-ed rendering code on a considerably slower computer, I did not believe that this alone could explain the performance difference against the old code which rendered a page in about 1 second. Since the only real change was the move from Jinja macros to JinjaX components, I suspected JinjaX to be responsible. A rewrite of two "hot paths" away from JinjaX reduced the rendering time from 12 to about 4 seconds. This clearly pointed to a JinjaX-related issue.

To understand what was going on, I created a simple benchmark (see the attached archive). The benchmarking code renders a simple list-like HTML structure where each element of the list is rendered by a Jinja macro or a JinjaX component, respectively. Total time needed to create Jinja environment, JinjaX Catalog and to render the template is measured.

Plain Jinja performs like this:
```
$> ./main.py --loops 50000
$> Total time needed to run the render function: 0.22861504554748535 sec
```

This is a staggering difference compared to JinjaX
```
$> ./main.py --loops 50000 --jinjax
$> Total time needed to run the render function:  10.462630271911621 sec
```

This makes JinjaX about 45x slower than Jinja in this simple test.

Each JinjaX component must be backed by a separate file. When a JinjaX component is rendered, [`catalog.render()`](https://github.com/jpsca/jinjax/blob/9951b5d708151e031e9d4aa232140685ab038827/src/jinjax/catalog.py#L115) function is called for each "instance" of that component on a page. As one of the first steps, JinjaX reads the content of the backing file from disk. There is no caching involved anywhere so if a component is present 1000 times on a page, it's file will be `open()`ed, `stat()`ed, `read()` and `close()`d that many times. A crude hack to allow for caching of component files helped a bit:

```
$> ./main.py --loops 50000 --jinjax
$> Total time needed to run the render function:  8.518791198730469 sec
```

Another bottleneck is caused by JinjaX needing to calculate the path to each component file [every time it renders it](https://github.com/jpsca/jinjax/blob/9951b5d708151e031e9d4aa232140685ab038827/src/jinjax/catalog.py#L134). The [`_get_component_path()`](https://github.com/jpsca/jinjax/blob/9951b5d708151e031e9d4aa232140685ab038827/src/jinjax/catalog.py#L220) queries the filesystem in a non-trivial loop, again making an unknown number of syscalls. Another crude hack to cache the paths showed that this constant path calculation is an even bigger problem than the constant re-reading of the component files.

```
$> ./main.py --loops 50000 --jinjax
$> Total time needed to run the render function:  4.333554983139038 sec
```

After that, there are no easy performance wins to be made. Jinja is fast because - as far as I understand - it compiles its templates to some kind of bytecode to avoid the expensive parsing. JinjaX appears to reparse components on every re-render.

[jinjax_bench.zip](https://github.com/Molecular-Biophysics-Database/mbdb-app/files/12565631/jinjax_bench.zip)

**Comment by edager** [comment]
Thanks for doing the research on this!

Although this might be solved by using a bigger hammer in deployment (i.e.  CESNET infrastructure vs. PC hosting), it will be rather painful from development perspective I would say, as Biocev can't get admin rights so we basically have to run it locally for development.  

@mirekys and @Alzpeta, do you have ideas how it could be improved or is this the price for switching to jinjaX?

**Comment by mirekys** [comment]
Thanks @MadCatX for raising this issue and for your work on on benchmarking & profiling! You probably already have more knowledge into this topic than the two of us, I'm afraid :). I think this needs to be brought to attention of the library authors, they might have some ideas on how/where to improve rendering (if not, we might need to seriously consider falling back to original macros, atleast for complex pages like this, or we will need to dedicate some serious effort on deep understanding of the library & how it could be improved). Could you please raise an issue [there](https://github.com/jpsca/jinjax/issues/new) & provide them with the benchmark? Even this info here could already help them to try to improve the library for everyone. Thanks a lot 👍

**Comment by mesemus** [comment]
This can be imho easily fixed on our side just with a simple caching (in production, not when flask_debug is set). @Alzpeta - please have a look at it.

**Comment by MadCatX** [comment]
@mirekys Considering the activity on the JinjaX's GH repo, I appears than JinX is a project of a single person developed in spare time rather than a systematically developed thing. Therefore I'm afraid that we can't realistically expect much to be done in the performance front. If you'd like to raise an issue with JinX developer, by all means do so but I'm afraid that I personally won't have time to follow up on this.

@mesemus Caching may alleviate the worst of it but even the developer experience is very much affected by JinX. The reason why I noticed the problem in the first place was because every single refresh of the record view was taking noticeably more and more time as I was adding more items for display.

In mbdb-app I solved the immediate issue by going back to plain Jinja macros. Only the top-level components remain as JinX components.

I think we can close this as "investigated/known".
